exercises/dominoes/test/Tests.hs

```diff
@@ -0,0 +1,108 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE RecordWildCards #-}
+
+import Control.Monad     (forM_, unless)
+import Data.Foldable     (for_)
+import Data.Function     (on)
+import Test.Hspec        (Spec, describe, expectationFailure, it, shouldBe, shouldMatchList)
+import Test.Hspec.Runner (configFastFail, defaultConfig, hspecWith)
+import Text.Printf       (printf)
+
+import Dominoes (chain)
+
+main :: IO ()
+main = hspecWith defaultConfig {configFastFail = True} specs
+
+specs :: Spec
+specs = describe "dominoes" $
+          describe "chain" $ for_ cases test
+  where
+
+    test Case{..} = it description assertion
+      where
+        assertion = if expected
+                      then shouldBeChain $ chain input
+                      else chain input `shouldBe` Nothing
+
+        shouldBeChain Nothing = expectationFailure "should have had a chain, but didn't"
+        shouldBeChain (Just output) = do
+          output `shouldMatchDominoesOf` input
+          consecutivesShouldMatch output
+          endsShouldMatch output
+
+        shouldMatchDominoesOf = shouldMatchList `on` map sortDomino
+
+        consecutivesShouldMatch l = forM_ (indexedPairs l) $ \(d1, d2, i) ->
+          shouldBeConsecutive l ("domino " ++ show i, d1) ("domino " ++ show (i + 1), d2)
+
+        endsShouldMatch [] = return ()
+        endsShouldMatch l@(d1:_) =
+          shouldBeConsecutive l (" last domino", last l) ("first domino", d1)
+
+        indexedPairs :: [a] -> [(a, a, Int)]
+        indexedPairs l = zip3 l (drop 1 l) [1..]
+
+        sortDomino (a, b) = if a > b then (b, a) else (a, b)
```

Comment:
I really tried to find something more readable without `if...then...else`, but the best I could do was:

```haskell
sortDomino = on (&&&) uncurry min max
```

If feel that maybe there a type class for homogeneous tuple allowing us to sort that, but this has nothing to do with this PR. :+1: